addons/mail/static/src/components/message_list/message_list.js

```diff
@@ -537,6 +552,12 @@ class MessageList extends Component {
             scrollTop,
             threadViewer: this.threadView.threadViewer,
         });
+        // Automatically scroll to new received messages only when the list is
+        // currently fully scrolled.
+        const hasAutoScrollOnMessageReceived = (this.props.order === 'asc')
+            ? scrollTop === this.el.scrollHeight - this.el.clientHeight
+            : scrollTop === 0;
+        this.threadView.update({ hasAutoScrollOnMessageReceived });
```

Comment:
Being precisely at top/bottom is not great. Old code had offset of 20 pixels to improve that.

Though the better code should have new message auto-scroll based on (partial) visibility of the last message before reception of message. As long as this decision is made at component-level, it cannot be much improved.

Reply:
I think the twitch approach is best to this: there should be a visual to indicate whether auto-scroll is enabled or not "restore auto scroll" which goes to the end of list immediately + restore auto feature.

In this context I think being exactly at the bottom is fine, I wouldn't change that here, though if there is a strong argument its just as easy to just add some extra px in those conditions.